File: Executables/Examples/example_MCMC.cpp
#include <chrono>
#include "NSL.hpp"
#include "highfive/H5File.hpp"
#include <yaml-cpp/yaml.h>

int main(int argc, char* argv[]){
    YAML::Node system;
    NSL::Logger::init_logger(argc, argv);

    system = YAML::LoadFile(argv[1]);

    // Define the device to run on NSL::GPU(ID=0) or NSL::CPU(ID=0)
    auto device = NSL::CPU();

    auto init_time =  NSL::Logger::start_profile("Program Initialization");

    // Define the parameters of your system (you can also read these in...)
    typedef NSL::complex<double> Type;

    // Typically you want to read these in or provide as an argument to such 
    // a code but for this example we just specify them here
    // System Parameters
    //    Inverse temperature 
    Type beta = system["beta"].as<double>();
    
    //    On-Site Coupling
    Type U    = system["U"].as<double>();

    //    Number of time slices
    NSL::size_t Nt = system["nt"].as<int>();
    
    std::string H5NAME(
       fmt::format("./"+system["h5file"].as<std::string>())
    );  // name of h5 file to store configurations, measurements, etc. . .
    NSL::H5IO h5(H5NAME, NSL::File::Truncate);
    std::string BASENODE(
       fmt::format(system["name"].as<std::string>())
    );

    // Leapfrog Parameters
    //      Trajectory Length
    NSL::RealTypeOf<Type> trajectoryLength = system["trajectory length"].as<double>(); // We ensure that this is a real number in case Type is complex
    //      Number of Molecular Dynamics steps
    NSL::size_t numberMDsteps = system["MD steps"].as<int>();
    
    // Markov Change Parameters 
    //     Number of Burn In configurations to thermalize the chain
    NSL::size_t NburnIn = 1000;
    //     Number of configurations to be computed on which we will measure
    NSL::size_t Nconf = 500;
    //     Number of configurations not used for measurements in between each stored configuration
    NSL::size_t saveFreq = system["checkpointing"].as<int>();
    // The total number of configurations is given by the product:
    // Nconf_total = Nconf * saveFreq

    // Define the lattice geometry of interest
    // ToDo: Required for more sophisticated actions
    NSL::Lattice::Generic<Type> lattice(system);
    NSL::size_t dim = lattice.sites();

    // Put the lattice on the device. (copy to GPU)
    lattice.to(device);

    // write out the physical and run parameters for this system
    HighFive::File h5file = h5.getFile();

    // lattice name
    HighFive::DataSet dataset = h5file.createDataSet<std::string>(BASENODE+"/Meta/lattice",HighFive::DataSpace::From(lattice.name()));
    dataset.write(lattice.name());

    // U
    dataset = h5file.createDataSet<std::complex<NSL::RealTypeOf<Type>>>(BASENODE+"/Meta/params/U", HighFive::DataSpace::From(static_cast <std::complex<NSL::RealTypeOf<Type>>> (U)));
    dataset.write(static_cast <std::complex<NSL::RealTypeOf<Type>>> (U));
        
    // beta
    dataset = h5file.createDataSet<std::complex<NSL::RealTypeOf<Type>>>(BASENODE+"/Meta/params/beta",HighFive::DataSpace::From(static_cast <std::complex<NSL::RealTypeOf<Type>>> (beta)));
    dataset.write(static_cast <std::complex<NSL::RealTypeOf<Type>>> (beta));

    // Nt
    dataset = h5file.createDataSet<NSL::size_t>(BASENODE+"/Meta/params/Nt",HighFive::DataSpace::From(Nt));
    dataset.write(Nt);

    // dim
    dataset = h5file.createDataSet<NSL::size_t>(BASENODE+"/Meta/params/spatialDim",HighFive::DataSpace::From(dim));
    dataset.write(dim);

    // Nmd
    dataset = h5file.createDataSet<NSL::size_t>(BASENODE+"/Meta/params/nMD",HighFive::DataSpace::From(numberMDsteps));
    dataset.write(numberMDsteps);

    // saveFreq
    dataset = h5file.createDataSet<NSL::size_t>(BASENODE+"/Meta/params/saveFreq",HighFive::DataSpace::From(saveFreq));
    dataset.write(saveFreq);

    // trajectory length
    dataset = h5file.createDataSet<NSL::RealTypeOf<Type>>(BASENODE+"/Meta/params/trajLength",HighFive::DataSpace::From(trajectoryLength));
    dataset.write(trajectoryLength);

    // action type
    std::string action = "hubbardExp";
    dataset = h5file.createDataSet<std::string>(BASENODE+"/Meta/action",HighFive::DataSpace::From(action));
    dataset.write(action);

    // get number of ions
    NSL::size_t Nx = lattice.sites();

    NSL::Logger::info("Setting up a Hubbard action with beta={}, Nt={}, U={}, on a {}.", NSL::real(beta), Nt, NSL::real(U), lattice.name());

    // Put the action parameters into the appropriate container
    NSL::Action::HubbardGaugeAction<Type>::Parameters params(
        /*beta=*/  beta,
        /*Nt = */  Nt,    
        /*U =  */  U
    );
    NSL::Action::HubbardFermionAction<Type,decltype(lattice),
        NSL::FermionMatrix::HubbardExp<Type,decltype(lattice)>>::Parameters paramsHFM(
        /*beta=*/  beta,
        /*Nt = */  Nt,    
        /*lattice=*/lattice
    );

    NSL::Action::HubbardGaugeAction<Type> S_gauge(params);
    NSL::Action::HubbardFermionAction<Type,decltype(lattice),NSL::FermionMatrix::HubbardExp<Type,decltype(lattice)>> S_fermion(paramsHFM);

    // Initialize the action
    NSL::Action::Action S = S_gauge + S_fermion;

    // Initialize a configuration as starting point for the MC change
    // For CPU code put here
    NSL::Configuration<Type> config{
        {"phi", NSL::Tensor<Type>(device,Nt,Nx)}
    };
    config["phi"].randn();
    config["phi"].imag() = 0;
    config["phi"] *= params.Utilde;

    NSL::Logger::info("Setting up a leapfrog integrator with trajectory length {} and {} MD steps.", trajectoryLength, numberMDsteps);

    // Initialize the integrator
    NSL::Integrator::Leapfrog leapfrog( 
        /*action*/S,  
        /*trajectoryLength*/trajectoryLength,
        /**numberSteps*/numberMDsteps
    );

    // Initialize the HMC
    NSL::MCMC::HMC hmc(leapfrog, S, h5);
    NSL::Logger::stop_profile(init_time);

    // Burn In
    // We can pass just a config to the generate function a MarkovState is 
    // generated automatically. If we want more control you cane also provide
    // a MarkovState.
    // The Template argument Chain{AllStates,LastState} is a memory optimization
    // where the LastState will return only the last generated state and does 
    // not keep the rest in memory while the AllStates (see production for use)
    // will store all states according to the saveFrequency.

    auto therm_time =  NSL::Logger::start_profile("Thermalization");
    NSL::Logger::info("Thermalizing {} steps...", NburnIn);
    NSL::MCMC::MarkovState<Type> start_state = hmc.generate<NSL::MCMC::Chain::LastState>(config, NburnIn);
    NSL::Logger::stop_profile(therm_time);

    // Generate Markov Chain
    // Here we should store the entire chain (i.e. every saveFreq element)
    // This generates Nconf*saveFreq configurations, though the std::vector
    // which is returned is of length Nconf.
    // 
    // Note: This also has a overload for providing a configuration only.
    auto gen_time =  NSL::Logger::start_profile("Generation");
    NSL::Logger::info("Generating {} steps, saving every {}...", Nconf, saveFreq);
    std::vector<NSL::MCMC::MarkovState<Type>> markovChain = hmc.generate<NSL::MCMC::Chain::AllStates>(start_state, Nconf, saveFreq, BASENODE+"/markovChain");
    NSL::Logger::stop_profile(gen_time);

    // Print some final statistics
    NSL::Logger::info("Acceptance Rate: {}%", NSL::MCMC::getAcceptanceRate(markovChain) * 100);

    return EXIT_SUCCESS;
}

File: Python/example_MCMC.py
# cmake creates a symlink of this script for development purposes. Therefore, the directory of the symlink needs to be identified for some functionality.
import os
import sys
symlink_dir = os.path.dirname(os.path.abspath(__file__))
# add symlink directory to path for PyNSL import
sys.path.insert(0, symlink_dir)

import PyNSL as nsl
import yaml
import h5py as h5
import torch
import numpy as np

def main():

    # Example of how to read parameters from a yaml file
    # with open('../Executables/Examples/example_param.yml', 'r') as f:
        # yml = yaml.safe_load(f)

    # Python dictionary containing HMC parameters
    params = {
        'name': "two_sites",
        'Nx': 2,
        'Nt': 16,
        'beta': 10.,
        'U': 2.,
        'mu': 0.0,
        'offset': 0.0,
        'save frequency': 2,
        'Ntherm': 100,
        'Nconf': 100,
        'trajectory length': 2.,
        'Nmd': 5,
        'h5file': "example.hmc.h5",
    }

    # Set torch device to GPU if available
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    print(f"Using {device} device.")

    # Print parameters
    for key, value in params.items():
        print(f"{key:20} {value}")
    
    # Create lattice from yaml file
    lattice = nsl.Lattice.Generic('../Executables/Examples/example_param.yml')
    Nx = lattice.sites()
    # Move lattice to device
    lattice.to(device)

    # Initialize h5 file and write metadata
    with h5.File(params['h5file'], 'w') as h5f:
        basenode = h5f.create_group(params['name'])
        write_meta(params, basenode)

    # Create Hubbardmodel action as sum of gauge and fermion action
    action = nsl.Action.SumAction(nsl.Action.HubbardGaugeAction(params), nsl.Action.HubbardFermionAction(lattice, params))
    
    # Create starting field
    init_phi = torch.randn((params['Nt'], Nx), device=torch.device(device))
    init_phi = torch.tensor(init_phi, dtype=torch.complex128) + 1.j*params['offset']
    print(init_phi)
    # Set initial configuration
    init_config = {"phi": init_phi}
    
    # Create real valued Leapfrog integrator and HMC
    lfr = nsl.Integrator.LeapfrogRealForce(action, params['trajectory length'].real, params['Nmd'])
    hmc = nsl.MCMC.HMC(lfr, action, params['h5file'])
    
    # Create initial Markov state from initial configuration
    init_state = nsl.MCMC.MarkovState(init_config, action(init_config), 1.)
    
    # Thermalize (i.e. generate beginning of Markov chain but do not save configurations)
    thermalized_state = hmc.thermalize(init_state, params['Ntherm'], params['save frequency'])

    # Generate Markov chain saving Nconf configurations with frequency saveFrequency
    mc = hmc.generate(thermalized_state, Nconf=params['Nconf'], saveFrequency=params['save frequency'])

    # Print field at position 42 of the Markov chain as sanity check
    print(mc[42].configuration["phi"])

def write_meta(params, basenode):
    basenode["Meta/lattice"] = str(params["name"])

    basenode["Meta/params/U"] = complex(params["U"])

    basenode["Meta/params/beta"] = complex(params["beta"])

    basenode["Meta/params/mu"] = complex(params["mu"])

    basenode["Meta/params/Nt"] = np.uint64(params["Nt"])

    basenode["Meta/params/spatialDim"] = np.uint64(params["Nx"])

    basenode["Meta/params/nMD"] = np.uint64(params["Nmd"])

    basenode["Meta/params/saveFreq"] = np.uint64(params["save frequency"])

    basenode["Meta/params/trajLength"] = float(params["trajectory length"])

    action = "hubbardExp"
    # h5file[basenode + "/Meta/action"] = np.string_(action)

if __name__ == '__main__':
    main()